src/gccrs/args.rs

```diff
@@ -0,0 +1,152 @@
+//! This module interprets arguments given to `rustc` and transforms them into valid
+//! arguments for `gccrs`.
+
+use getopts::{Matches, Options};
+
+/// Crate types supported by `gccrs`
+#[derive(Clone, Copy)]
+pub enum CrateType {
+    /// Binary application
+    Bin,
+    /// Dynamic library/Shared object
+    DyLib,
+    /// Statically linked library
+    StaticLib,
+    /// Remaining options, handled by `rustc` but not `gccrs`
+    Unknown,
+}
+
+impl CrateType {
+    /// Get the corresponding [`CrateType`] from a given option to the `--crate-type`
+    /// option
+    pub fn from_str(s: &str) -> CrateType {
+        match s {
+            "bin" => CrateType::Bin,
+            "dylib" => CrateType::DyLib,
+            "staticlib" => CrateType::StaticLib,
+            _ => CrateType::Unknown,
+        }
+    }
+}
+
+/// Structure used to represent arguments passed to `gccrs`. Convert them from `rustc`
+/// arguments using [`GccrsArg::from_rustc_arg`]
+pub struct GccrsArgs {
+    source_files: Vec<String>,
+    crate_type: CrateType,
+    output_file: String,
+}
+
+impl GccrsArgs {
+    fn generate_parser() -> Options {
+        let mut options = Options::new();
+        options.optopt("", "crate-name", "Name of the crate to compile", "NAME");
+        options.optopt("", "edition", "Rust edition to use", "YEAR");
+        options.optopt("", "error-format", "Requested error format", "EXTENSION");
+        options.optopt(
+            "",
+            "out-dir",
+            "Directory in which to output generated files",
+            "DIR",
+        );
+        options.optopt("", "emit", "Requested output to emit", "KIND");
+        options.optopt("", "json", "JSON Rendering type", "RENDER");
+        options.optopt("", "json", "JSON Rendering type", "RENDER");
+        options.optmulti("C", "", "Extra compiler options", "OPTION[=VALUE]");
+        options.optmulti(
+            "L",
+            "",
+            "Add a directory to the library's search path",
+            "KIND[=PATH]",
+        );
+        options.optmulti("", "crate-type", "Type of binary to output", "TYPE");
+
+        options
+    }
+
+    fn format_output_filename(matches: &Matches, crate_type: &CrateType) -> String {
+        // FIXME: No unwraps here
+        let crate_name = matches.opt_str("crate-name").unwrap();
+        let out_dir = matches.opt_str("out-dir").unwrap();
+
+        // FIXME: Figure out a way to return multiple output filenames. Just handle `bin`
+        // for now
+        let c_options = matches.opt_strs("C");
+
+        // FIXME: This is probably different on Windows, we should use Paths and PathStrs
+        // instead
```

Comment:
Yes, use `Path[Buf]` everywhere you can. This will save you from many headaches.